[omatic] Initial contribution

seaside1 opened the pull request:
State O-Matic Binding

[omatic] Initial Contribution 


DESCRIPTION

The State O-Matic Binding will enable you to monitor electrical devices such as washing machine, dryer , dishwasher, electrical car charging and so on. The component is generic and any type of equipment can be monitored as long as it is possible to read its power consumption.

Examples of usage:

    Monitor your electric car. Once the car has finished charging, you will get notified about the cost, consumption and time it took.
    Monitor your dishwasher, dryer or washing machine. Once they are finished you will get notified about the cost, consumption and time it took.
    Send info when the power consumption is above a certain limit for a device.

TESTING 

There are some basic junit tests.
Jar file for OpenHAB 3.0 can be downloaded on: https://github.com/seaside1/omatic/tree/master/bin

See this thread for more info: https://community.openhab.org/t/state-o-matic-binding-monitor-your-washingmachine-dishwasher-electric-car-etc/104494

seaside1 commented:
@fwolter  Can we trigger  a new build? From Jenkins I see this:

[PR-openHAB-Addons@3] $ /home/jenkins/jenkins-agent1/tools/hudson.model.JDK/OpenJDK_11/bin/java -Xms1g -Xmx2g -Djava.awt.headless=true -cp /home/jenkins/jenkins-agent1/maven35-agent.jar:/home/jenkins/jenkins-agent1/tools/hudson.tasks.Maven_MavenInstallation/maven_latest_/boot/plexus-classworlds-2.6.0.jar:/home/jenkins/jenkins-agent1/tools/hudson.tasks.Maven_MavenInstallation/maven_latest_/conf/logging jenkins.maven3.agent.Maven35Main /home/jenkins/jenkins-agent1/tools/hudson.tasks.Maven_MavenInstallation/maven_latest_ /usr/share/jenkins/agent.jar /home/jenkins/jenkins-agent1/maven35-interceptor.jar /home/jenkins/jenkins-agent1/maven3-interceptor-commons.jar 39851
ERROR: Aborted Maven execution for InterruptedIOException
java.net.SocketTimeoutException: Accept timed out
	at java.net.PlainSocketImpl.socketAccept(Native Method)

seaside1 commented:
@wborn 
Seems like the Mqtt binding failed:

org.openhab.binding.mqtt.handler.BrokerHandlerTest.handlerInit
https://ci.openhab.org/job/PR-openHAB-Addons/org.openhab.addons.bundles$org.openhab.binding.mqtt/1250/testReport/junit/org.openhab.binding.mqtt.handler/BrokerHandlerTest/handlerInit/

Retrigger again? :)

wborn commented:
Maybe we're more lucky this time. :four_leaf_clover:  :stuck_out_tongue_winking_eye:

seaside1 commented:
Still no luck it seems. This time https://ci.openhab.org/job/PR-openHAB-Addons/1251/org.openhab.addons.itests$org.openhab.binding.feed.tests/

Feed binding tests failed.

wborn commented:
There's even an issue for it https://github.com/openhab/openhab-addons/issues/8820.

seaside1 commented:
> There's even an issue for it #8820.

Too bad. Please trigger again and I'll keep all my fingers crossed.🤞

seaside1 commented:
@wborn , @fwolter 
I finally got the binding to build successfully, I rebased against main to get the latest updates to the integration test of some of the other biding.

fwolter commented on bundles/org.openhab.binding.omatic/LICENSE:
```diff
@@ -0,0 +1,277 @@
+Eclipse Public License - v 2.0
```
The legal things are already covered by the main project and the header in each of your files. So, this could be removed.

seaside1 replied:
Fixed in https://github.com/openhab/openhab-addons/pull/9286/commits/ebd0ae674be0e8107722de21ba936c75db3d6f20

fwolter commented on bundles/org.openhab.binding.omatic/README.md:
```diff
@@ -0,0 +1,219 @@
+# State O-Matic Binding
+The State O-Matic Binding will enable you to monitor electrical devices such as
+washing machine, dryer , dishwasher, electrical car charging and so on.
```
```suggestion
washing machine, dryer, dishwasher, electrical car charging and so on.
```

fwolter commented on bundles/org.openhab.binding.omatic/README.md:
```diff
@@ -0,0 +1,219 @@
+
+<img src="https://raw.githubusercontent.com/seaside1/omatic/master/image/omatic.jpg" width=200 />
+
+Please take a look at ThomDietrich post in OpenHAB forums about monitoring your washingmachine https://community.openhab.org/t/washing-machine-state-machine/15587
```
```suggestion
Please take a look at ThomDietrich's [post](https://community.openhab.org/t/washing-machine-state-machine/15587) in the openHAB forum about monitoring your washing machine.
```

fwolter commented on bundles/org.openhab.binding.omatic/README.md:
```diff
@@ -0,0 +1,219 @@
+- Monitor your dishwasher, dryer or washing machine. Once they are finished you will get notified about the cost, consumption and time it took.
+- Send info when the power consumption is above a certain limit for a device.
+
+<img src="https://raw.githubusercontent.com/seaside1/omatic/master/image/omatic.jpg" width=200 />
```
You could include images in the `doc/` folder of this binding.

seaside1 replied:
Fixed in https://github.com/openhab/openhab-addons/pull/9286/commits/ebd0ae674be0e8107722de21ba936c75db3d6f20

fwolter reviewed (commented):
This is a very interesting contribution. I'd love to see this in openHAB! Did you take a look at the new automation engine API? You can implement modules with inputs and outputs and use those in the rule engine easily. I'm currently implementing a PID controller on that basis. You could take a look at https://github.com/fwolter/openhab-addons/tree/pidcontroller/bundles/org.openhab.action.pidcontroller/src/main/java/org/openhab/action/pidcontroller/internal for some inspiration.

seaside1 commented:
@fwolter Big thank you for the feedback. I'll make some adjustments and get back to you.  I'll checkout the automation engine API it looks interesting.

Regards S

seaside1 commented:
I've done the fixes. I'll take a look at the automation API as well.

seaside1 commented:
@fwolter 
I've removed the bridging of items. The binding now utilizes the ItemRegistry and listens to the eventbus.
I found that this will make it a lot simpler for the user, no need for DSL rules. I believe this is a better solution than using the automation api.

Also rebased against main to make sure it can be merged, whenever it is ready.

seaside1 commented:
And I did end up in a strange state with the signatures
Commit sha: af3e310, Author: jenkins, Committer: Joseph Hagberg; jenkins@8e1733d9c8f4 is not a valid email address.


git commit --amend --signoff

Now your commits will have your sign off. Next run

git push --force-with-lease origin main

Does not seem to work. Any ideas?

seaside1 commented:
@fwolter  , @wborn 

This last rebase I did really went south. Is it ok if I close this PR? I'll reopen a new one against 3.1 instead and link to this closed one.

fwolter commented:
It seems the easiest way to open a new PR.

But before opening a new PR, some general considerations. Bindings are intended to interface other systems. I understand that your binding takes the values from Items, do some calculations and output the result on another Item. This use case fits exactly into the rule engine and the automation API. Nowadays the user doesn't need to write DSL rules, but can configure the settings (Items) via the UI. See this [screenshot ](https://community-openhab-org.s3.dualstack.eu-central-1.amazonaws.com/optimized/3X/a/9/a9aaa73ea4b4b396857285253bf0e90b117edce4_2_694x1000.png) for example. It's the configuration interface for the [PID controller](https://github.com/fwolter/openhab-addons/tree/pidcontroller/bundles/org.openhab.automation.pidcontroller). It takes Items, do some math and output the result on another Item. This could be the perfect template for your module.

seaside1 commented:
> It seems the easiest way to open a new PR.
> 
> But before opening a new PR, some general considerations. Bindings are intended to interface other systems. I understand that your binding takes the values from Items, do some calculations and output the result on another Item. This use case fits exactly into the rule engine and the automation API. 

I have tested the PID Controller. I agree that this might be better for the average user. I'll reopen a new PR when / if I implement this. 

Regards, S
